classes/file_search.php

```diff
@@ -616,6 +646,79 @@ public static function make_where_clause(object $criteria): array {
 
         return [$whereclause, $params];
     }
+
+    /**
+     * Determine the id of a line from the output csv file.
+     *
+     * @param string $line of an output file.
+     */
+    public static function get_id_from_csv(string $line): int {
+        // Interpret the last line as a csv line.
+        $csv = str_getcsv($line);
+        // Check a few columns to ensure we have a valid line.
+        if (empty($csv[self::CSV_CONTEXTID])) {
+            return 0;
+        }
+        if ( ! ctype_digit($csv[self::CSV_CONTEXTID])) {
+            return 0;
+        }
+        if (empty($csv[self::CSV_FILEID])) {
+            return 0;
+        }
+        if ( ! ctype_digit($csv[self::CSV_FILEID])) {
+            return 0;
+        }
+
+        return $csv[self::CSV_FILEID];
+    }
+
+    /**
+     * Look at the previous output file to decide how to resume.
+     *
+     * @param string $filename
+     * @return int $resumeid The first id that should be scanned.
+     * @return int $matchcount The number of matches left in the file.
+     */
+    public static function resume(string $filename): array {
+        if ( ! file_exists($filename)) {
+            return [0, 0];
+        }
+
+        $lines = file($filename, FILE_IGNORE_NEW_LINES);
+        if (count($lines) < 3) {
+            // Too small to resume.
+            return [0, 0];
+        }
+        $lastline = end($lines);
+        $resumeid = self::get_id_from_csv($lastline);
+        if (empty($resumeid)) {
+            array_pop($lines);
+            $lastline = end($lines);
+            $resumeid = self::get_id_from_csv($lastline);
+            if (empty($resumeid)) {
+                // If last two lines are bad, give up.
+                return [0, 0];
+            }
+        }
+        // Now remove all lines that match this id.
+        while (true) {
+            array_pop($lines);
+            $line = end($lines);
+            $lineid = self::get_id_from_csv($line);
+            if ($lineid != $resumeid) {
+                // Leave this line in place.
+                break;
+            }
+            if ( count($lines) < 3 ) {
+                // Too small to resume.
+                return [0, 0];
+            }
+        }
+        // Re-write the file, without the matched id lines.
+        file_put_contents($filename, implode("\n", $lines) . "\n");
```

Comment:
If this is running on a non linux system would the original lines be split differently? If so would PHP_EOL be the better option?

Reply:
OK. I will fix this.